### PR TITLE
Fixes #19146: in some case inverting a request leads to an error and 0 results

### PR DESCRIPTION
**Pull request by fanf**
https://issues.rudder.io/issues/19146

### CODE REVIEW

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala [hunk 1]
```diff
@@ -414,58 +424,51 @@ class InternalLDAPQueryProcessor(
...
+            fss     = filterSeqSet(dms)
+            blf     <- buildLastFilter(nq, fss)
+
+            // for convenience
+            (finalLdapFilter, finalSpecialFilters) = blf
+
+            //final query, add "match only server id" filter if needed
+            rt      = nodeObjectTypes.copy(filter = finalLdapFilter)
+            _       <- logPure.debug(s"[${debugId}] |- (final query) ${rt}")
+            entries <- for {
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:447]: We don't need the for here if i read correctly

Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:447]: right, it was simplified enought :)

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala [hunk 2]
```diff
@@ -414,58 +424,51 @@ class InternalLDAPQueryProcessor(
...
       }
+      inverted <- query.transform match {
+                  case ResultTransformation.Identity => results.succeed
+                  case ResultTransformation.Invert   =>
+                    for {
+                      _      <- logPure.debug(s"[${debugId}] |- (need to get all nodeIds for inversion) ")
+                      allIds <- executeQuery(nodeObjectTypes.baseDn, nodeObjectTypes.scope, nodeObjectTypes.objectFilter, Some(ALL), Set(), Set(), nq.composition, debugId)
+                      ids    =  results.flatMap(x => x(A_NODE_UUID)).toSet
+                      res    =  allIds.collect { case e if(!ids.contains(e.value_!(A_NODE_UUID))) => e }
+                      _      <- logPure.debug(s"[${debugId}] |- (invert) entries after inversion: ${res.size}")
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:464]: why a collect and not a filter or a filterNot ?

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:464]: ie allIds.filterNot(e => ids.contains(e.value_!(A_NODE_UUID)))

Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:464]: I didn't change that. I imagine it used to be comething else in the right side of the `=>`

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:464]: That was added in a previous PR when you added the invert logic but it did not botherer me first time, and here yes, but that's not important

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala [hunk 3]
```diff
@@ -350,20 +350,30 @@ class InternalLDAPQueryProcessor(
...
+            case (Some(a),b) =>  Some(a union b._2 )
+            case (None,b)    =>  Some(b._2)
+          }
           case And =>
-            val s = if(dnMapSet.isEmpty) Set[DN]() else dnMapSet.foldLeft(dnMapSet.head._2)( _ intersect _._2 )
-            // Here if s is empty, it means we are ANDing with an empty Set, so we could simply drop all computation from there
-            if(s.isEmpty && normalizedQuery.transform != ResultTransformation.Invert) {
-              logPure.logEffect.debug(s"[${debugId}] `-> early stop query (non-inverted empty sub-query)")
-              return None
+            dnMapSet.toList match {
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:359]: i'm concerned about the convertion from map to list, and then from map to list to map again ( mapSet.toList.sequence.map(_.toMap) ) 
cost of convertion can be prohibitive with thousands of nodes

i don't have a better proposal though :/

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala:359]: I think i fixed it here in this PR: https://github.com/fanf/rudder/pull/13 

I never convert the original map, and only fold in an optionnal result map

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

**Review by VinceMacBuche — APPROVED** [review]

**Pull request merged**